### PR TITLE
MGMT-16199: block unbind host while cluster is progressing

### PR DESCRIPTION
**Pull request by eifrach**
## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by avishayt** [comment]
According to SourceStates, these transitions don't apply.  The bug isn't about unbind, it's about unregister.  If you look at UnRegisterHost() in internal/host/host.go, it just deletes the host from the DB without any checks.  We should add the check here (return 409).

**Comment by eifrach** [comment]
@avishayt  changed, should this solve it?

File: internal/host/host.go [hunk 1]
```diff
@@ -1441,14 +1441,28 @@ func (m *Manager) GetHostByKubeKey(key types.NamespacedName) (*common.Host, erro
...
+	}
+	var cluster common.Cluster
+	err := m.db.Select("status").Take(&cluster, "id = ?", h.ClusterID.String()).Error
+	if err != nil {
 		return err
 	}
-	if h.ClusterID != nil {
-		if err := m.db.Model(&common.Cluster{}).Where("id = ?", h.ClusterID).Update("trigger_monitor_timestamp", time.Now()).Error; err != nil {
-			return err
+	clusterProgressingStates := []string{models.ClusterStatusInstalling, models.ClusterStatusFinalizing}
```
Comment (gamli75) [inline internal/host/host.go:1452]: I think we should block it also in case of preparing-for-installation, @avishayt WDYT?

Comment (avishayt) [reply internal/host/host.go:1452]: Yes, and also installing-pending-user-action.  I suggest putting this list of states into internal/cluster/cluster.go, similar to CanDownloadFiles() or CanDownloadKubeconfig().

Another thing, while looking at that file, I see three places where we check for ClusterStatusInstalling but ignore ClusterStatusInstallingPendingUserAction:
1. CancelInstallation() doesn't allow canceling while in installing-pending-user-action
2. IsOperatorAvailable() will return false while in installing-pending-user-action
3. SetVipsData() won't catch VIPs changing during installation

Comment (eifrach) [reply internal/host/host.go]: @avishayt so we can't move it to cluster package due to import loop issues
so I've create it in common

let me know if it's ok i will start working on the test's

File: internal/common/common.go [hunk 2]
```diff
@@ -432,6 +435,21 @@ func GetInventoryInterfaces(inventory string) (string, error) {
 	return inventory[interfacesLocation : interfacesLocation+endLocation], nil
 }
 
+// blocking host removal while cluster is progressing
+func CanUnbindhost(c *Cluster) (err error) {
+	clusterStatus := swag.StringValue(c.Status)
+	NotAllowedStatuses := []string{
+		models.ClusterStatusFinalizing,
```
Comment (avishayt) [inline internal/common/common.go:442]: Also preparing-for-installation

**Review by avishayt — COMMENTED** [review]
Looks OK.  There are a bunch of unrelated formatting changes here.  I think it's better to not commit them for the sake of clarity.  If there is some formatting convention you feel strongly about, we can add consider adding it to the linter.

**Comment by eifrach** [comment]
> Looks OK. There are a bunch of unrelated formatting changes here. I think it's better to not commit them for the sake of clarity. If there is some formatting convention you feel strongly about, we can add consider adding it to the linter.

So the reason is that I'm running `skipper make lint` before pushing 

the linter is also fixing where it can, not sure why I am the only one doing it :smiley: 

I'll recreate the PR

**Comment by eifrach** [comment]
recreated #5809

/close

**Pull request closed without merging**